### PR TITLE
Greatly speed up doctests by compiling compatible doctests in one file

### LINKED ISSUE #75341: Rustdoc should run all doctests in one binary
Raised by @Lokathor on discord:

> Doctests are slower to run than integration tests. Each individual doctest block runs as a unique binary/process when testing, but each entire integration test file runs as a single process with integration tests. So the test themselves aren't different speeds, but because you're building less bins and spawning less child processes it's faster to go through the integration tests.

I don't think there's any intrinsic reason the tests have to be their own process? The only issue is that many of them have `fn main`, which would have to be rewritten somehow. This could speed up CI times across the Rust ecosystem, including for the rust compiler itself.

Somewhat related to https://github.com/rust-lang/rust/issues/51228.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #75341.

I split the changes as much as possible to make it as easy as possible to review (even though it's a huge lot of code changes...).

The following tests are not included into the combined doctests:
 * `compile_fail`
 * `#![no_std]`
 * have invalid AST
 * `test_harness`
 * no capture
 * `--show-output` (because the output is nicer without the extra code surrounding it)

Everything else is merged into one file. If this one file fails to compile, we go back to the current strategy: compile each doctest separately.

Because of the `edition` codeblock attribute, I couldn't make them all into one file directly so I grouped them by edition, but it should be pretty anecdotic.

In case the users want a specific doctest to be opted-out from this doctest merge, I added the `standalone` codeblock attribute:

```rust
/// ```rust,standalone
/// // This doctest will be run in its own process!
/// ```
```

Now the interesting part, I ran it on a few crates and here are the results (with `cargo test --doc` to only include doctests):

| crate | nb doctests | before this PR | with this PR |
| - | - | - | - |
| sysinfo | 227 | 4.6s | 1.11s |
| geos | 157 | 3.95s | 0.45s |
| core | 4604 | 54.08s | 23.20s |
| std | 1147 | 13.88s | 8.82s |

r? @notriddle

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Seems like where there are enough tests, `libtest` simply blocks and does nothing. I'll split tests by chunks of ~200 tests at once to start and then check how to improve output.

**Comment by petrochenkov** [comment]
There should probably be some opt out from this merging.

There are numerous reasons why two pieces of code may cause issues when combined together (they should compete for some common resource, in general, and that is not always detectable by a compiler).
This was previously attempted with run-pass tests from the compiler test suite, but it ended up being too brittle.

**Comment by GuillaumeGomez** [comment]
It fails in multiple tests:

```
running 28 tests
test src/fmt/humantime.rs - fmt::humantime::Formatter::timestamp (line 17) ... ok
test src/fmt/mod.rs - fmt (line 21) ... ok
test src/fmt/mod.rs - fmt (line 42) ... ok
test src/fmt/mod.rs - fmt::Formatter (line 121) ... ok
test src/lib.rs - (line 175) ... ok
test src/lib.rs - (line 234) ... ok
test src/logger.rs - logger::Builder::filter (line 382) ... ok
test src/logger.rs - logger::Builder::filter_level (line 360) ... ok
test src/logger.rs - logger::Builder::format (line 231) ... ok
test src/logger.rs - logger::Builder::filter_module (line 341) ... ok
test src/logger.rs - logger::Builder::write_style (line 436) ... ok
test src/logger.rs - logger::Logger::from_default_env (line 600) ... ok
test src/logger.rs - logger::Builder::target (line 415) ... ok
test src/logger.rs - logger::Logger::from_env (line 567) ... ok
test src/logger.rs - logger::Logger::from_env (line 576) ... ok
test src/lib.rs - (line 253) ... FAILED
test src/lib.rs - (line 214) ... FAILED
test src/lib.rs - (line 19) ... FAILED
test src/logger.rs - logger::Builder (line 22) ... FAILED
test src/logger.rs - logger::Builder::from_default_env (line 176) ... FAILED
test src/logger.rs - logger::Builder::from_env (line 96) ... FAILED
test src/logger.rs - logger::Builder::from_env (line 86) ... FAILED
test src/logger.rs - logger::Builder::parse_default_env (line 199) ... FAILED
test src/logger.rs - logger::Builder::new (line 56) ... FAILED
test src/logger.rs - logger::Builder::parse_env (line 123) ... FAILED
test src/logger.rs - logger::Builder::parse_env (line 138) ... FAILED
test src/logger.rs - logger::init_from_env (line 927) ... FAILED
test src/logger.rs - logger::try_init_from_env (line 890) ... FAILED

failures:

---- src/lib.rs - (line 253) stdout ----
thread 'src/lib.rs - (line 253)' panicked at src/logger.rs:499:14:
Builder::init should not be called after logger initialized: SetLoggerError(())
stack backtrace:
   0:     0x564cf5d0b200 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h9cb9ba9de157d448
   1:     0x564cf5d3115b - core::fmt::write::h712ed17a4bea48d7
   2:     0x564cf5cf3ff9 - std::io::Write::write_fmt::ha9b4cf5497604558
   3:     0x564cf5d0afbe - std::sys_common::backtrace::print::h527405b0eb64d489
   4:     0x564cf5d0c09a - std::panicking::default_hook::{{closure}}::h7c8284f1aa6c5571
   5:     0x564cf5d0bdbf - std::panicking::default_hook::haf59e725ae5fe4cb
   6:     0x564cf5ca07fa - test::test_main::{{closure}}::h24de74550d40f595
   7:     0x564cf5d0c749 - std::panicking::rust_panic_with_hook::hc2212689298edd22
   8:     0x564cf5d0c4f6 - std::panicking::begin_panic_handler::{{closure}}::h2058e51712b5fdf6
   9:     0x564cf5d0b419 - std::sys_common::backtrace::__rust_end_short_backtrace::h1dc4a638a0e69753
  10:     0x564cf5d0c227 - rust_begin_unwind
  11:     0x564cf5b2dc16 - core::panicking::panic_fmt::h88caaa095c0a5b53
  12:     0x564cf5b2e586 - core::result::unwrap_failed::ha5ec7624a00a22e4
  13:     0x564cf5b391fe - core::result::Result<T,E>::expect::ha6de1769dcf31264
                               at /home/imperio/rust/rust/library/core/src/result.rs:1034:23
  14:     0x564cf5b3610d - env_logger::logger::Builder::init::h3eaa9af3496d3a3d
                               at /home/imperio/rust/env_logger/src/logger.rs:498:9
  15:     0x564cf5b31e74 - rust_out::__doctest_8::main::hc4539ff026f50669
  16:     0x564cf5b31ee3 - rust_out::__doctest_8::TEST::{{closure}}::hffdb7dfd4fa1d8d0
  17:     0x564cf5b30176 - core::ops::function::FnOnce::call_once::h74d127fff682a19d
  18:     0x564cf5ca5d92 - test::__rust_begin_short_backtrace::h377d4000c87befa2
  19:     0x564cf5cc18b2 - test::types::RunnableTest::run::h656e7414cfb3441c
  20:     0x564cf5ca6056 - test::run_test_in_process::hcb4c940c64146602
  21:     0x564cf5cae2bd - std::sys_common::backtrace::__rust_begin_short_backtrace::h6035574334cd54dd
  22:     0x564cf5cb89c5 - core::ops::function::FnOnce::call_once{{vtable.shim}}::hebcf07f42d5ca27a
  23:     0x564cf5d1757b - std::sys::pal::unix::thread::Thread::new::thread_start::h21bb0d8f9a9dc8fb
  24:     0x7fcbf7494ac3 - start_thread
                               at ./nptl/pthread_create.c:442:8
  25:     0x7fcbf7526850 - __GI___clone3
                               at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
  26:                0x0 - <unknown>
```

I implemented a fallback in case compilation fails. I could do the same for failing tests. Like that, we can support both current doctests with shared context and others who don't have have this shared context issue can enjoy having doctests running much faster. What do you think?

**Comment by GuillaumeGomez** [comment]
The logic currently is as follows: we merge tests, if the compilation of the combined tests failed, we fallback to the current behaviour.

One thing I thought though: we could spawn a new process for each test. That would prevent them from using "in program" common resources.

> This is also a problem with `TEST_BATCH_SIZE`, I think. If a test accidentally relies on a global resource set up by another test, then the tests might magically stop passing when 250 other tests get sandwiched between them.

Yes it's the current issue with `env_logger`, hence why I'm suggesting to either try running all tests in the batch independently or start a new process for each doctest.

**Comment by GuillaumeGomez** [comment]
What about changing the default way to run doctests starting the 2024 edition?

**Comment by GuillaumeGomez** [comment]
Agreed. Should I add the new `standalone` doctest attribute in this PR as well?

**Comment by GuillaumeGomez** [comment]
Once https://github.com/rust-lang/rust/pull/124561 is merged, I'll just remove the commit and rebase (one less commit, yeay!).

**Comment by GuillaumeGomez** [comment]
Seems like `x.py fmt` doesn't take into account files into `run-make`. Not convenient.

**Comment by jieyouxu** [comment]
> Seems like `x.py fmt` doesn't take into account files into `run-make`. Not convenient.

Files under `tests` are not formatted IIRC

**Comment by GuillaumeGomez** [comment]
Would you be ok if I added `run-make/rmake.rs` files into the `fmt` run?

**Comment by jieyouxu** [comment]
> Would you be ok if I added `run-make/rmake.rs` files into the `fmt` run?

That'd be perfectly reasonable yes

**Comment by GuillaumeGomez** [comment]
@notriddle: This is ready for review now.

File: tests/run-make/doctests-merge/doctest.rlib
Comment (jieyouxu) [inline tests/run-make/doctests-merge/doctest.rlib:1]: Is this file intended to be checked in? Maybe I'm missing something?

File: src/doc/rustdoc/src/write-documentation/documentation-tests.md [hunk 1]
```diff
@@ -376,6 +376,55 @@ that the code sample should be compiled using the respective edition of Rust.
...
+        assert!(!IS_INIT);
+        IS_INIT = true;
+    }
+}
+```
+
+If you run `rustdoc --test` on this code, it'll panic on the second doctest being
+run because `IS_INIT` value is not `false` anymore.
+
+This is where the `standalone` attribute comes in: ittells `rustdoc` that a doctest
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/write-documentation/documentation-tests.md:406]: Good catch!

File: src/librustdoc/doctest/tests.rs [hunk 2]
```diff
@@ -1,17 +1,38 @@
...
+        String::new(), // test name
+        LangString::empty_for_test(),
+        0,             // line
+        String::new(), // file name
+        Arc::new(IndividualTestOptions::empty()),
+        String::new(),  // test id
+        "",             // target_str
+        PathBuf::new(), // path
+        true,           // no_run
+    )
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/doctest/tests.rs:21]: That would indeed improve the code. :+1:

File: tests/run-make/doctests-merge/doctest.rlib
Comment (GuillaumeGomez) [reply tests/run-make/doctests-merge/doctest.rlib:1]: No it's not!

**Comment by GuillaumeGomez** [comment]
Rebased.

File: src/librustdoc/doctest.rs [hunk 3]
```diff
@@ -571,41 +610,367 @@ fn make_maybe_absolute_path(path: PathBuf) -> PathBuf {
...
+                    "\
+    fn main() {returns_result} {{
+        {}
+    }}",
+                    self.everything_else
+                )
+                .unwrap();
+            }
+        }
+        writeln!(
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/doctest.rs:886]: Ok, I'll add some extra checks for these cases. Good catch!

File: src/librustdoc/doctest.rs [hunk 4]
```diff
@@ -372,30 +371,47 @@ fn wrapped_rustc_command(rustc_wrappers: &[PathBuf], rustc_binary: &Path) -> Com
...
+        out_dir
+    };
+    if is_perm_dir && let Err(err) = std::fs::create_dir_all(&out_dir) {
+        eprintln!("Couldn't create directory for doctest executables: {err}");
+        panic::resume_unwind(Box::new(()));
+    }
+    out_dir.into()
+}
+
+fn run_test(
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/doctest.rs:398]: Oh I missed this one. Adding a struct.

File: src/librustdoc/doctest.rs [hunk 5]
```diff
@@ -994,8 +1330,231 @@ pub(crate) trait Tester {
...
+            Arc::clone(unused_externs),
+        ));
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct DocTestKinds {
+    /// Tests that cannot be run together with the rest (`compile_fail` and `test_harness`).
+    standalone: Vec<TestDescAndFn>,
+    others: FxHashMap<Edition, Vec<DocTest>>,
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/doctest.rs:1451]: With this PR, unless a doctest is "incompatible" (ie, has `compile_fail`, `test_harness`, `standalone`...), it is put into the according edition so all doctests from the same edition can be compiled together. We do not mix doctests from different editions. Hence why in here we have `Edition` as key whereas `standalone` here doesn't represent the code attribute but rather the tests that cannot (for sure) be run with the new combined doctest system.

File: tests/rustdoc-ui/doctest/failed-doctest-should-panic.rs [hunk 6]
```diff
@@ -1,7 +1,7 @@
 // FIXME: if/when the output of the test harness can be tested on its own, this test should be
 // adapted to use that, and that normalize line can go away
 
-//@ compile-flags:--test
+//@ compile-flags:--test -Z unstable-options --edition 2024
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply tests/rustdoc-ui/doctest/failed-doctest-should-panic.rs:4]: :+1:

File: src/doc/unstable-book/src/language-features/link-arg-attribute.md [hunk 7]
```diff
@@ -8,7 +8,7 @@ The `link_arg_attribute` feature allows passing arguments into the linker
 from inside of the source code. Order is preserved for link attributes as
 they were defined on a single extern block:
 
-```rust,no_run
+```rust,ignore (linking to "c" segfaults)
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/doc/unstable-book/src/language-features/link-arg-attribute.md:11]: I actually don't know how it could pass before. I opened #124043 about this.

File: src/doc/rustdoc/src/write-documentation/documentation-tests.md [hunk 8]
```diff
@@ -376,6 +376,55 @@ that the code sample should be compiled using the respective edition of Rust.
 # fn foo() {}
 ```
 
+Starting the 2024 edition, compatible doctests will be merged as one before being run.
+It means that they will share the process, so any change global/static variables will
+now impact the other doctests.
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/write-documentation/documentation-tests.md:381]: Much better indeed, thanks!

File: src/librustdoc/doctest.rs [hunk 9]
```diff
@@ -651,31 +1016,27 @@ pub(crate) fn make_test(
...
-                                Some(name) => found_extern_crate = name.as_str() == crate_name,
-                                None => found_extern_crate = item.ident.as_str() == crate_name,
+                        if let ast::ItemKind::ExternCrate(original) = item.kind {
+                            if !found_extern_crate && let Some(ref crate_name) = crate_name {
+                                found_extern_crate = match original {
+                                    Some(name) => name.as_str() == crate_name,
+                                    None => item.ident.as_str() == crate_name,
+                                };
                             }
                         }
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/doctest.rs:1033]: :+1:

File: src/librustdoc/doctest.rs [hunk 10]
```diff
@@ -636,7 +1001,7 @@ pub(crate) fn make_test(
             let dcx = DiagCtxt::new(Box::new(emitter)).disable_warnings();
             let psess = ParseSess::with_dcx(dcx, sm);
 
-            let mut found_main = false;
+            let mut found_main = None;
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/doctest.rs:1004]: :+1:

File: src/librustdoc/doctest.rs [hunk 11]
```diff
@@ -372,30 +371,47 @@ fn wrapped_rustc_command(rustc_wrappers: &[PathBuf], rustc_binary: &Path) -> Com
...
+}
+
+fn build_test_dir(outdir: &Arc<DirState>, is_multiple_tests: bool) -> PathBuf {
+    // Make sure we emit well-formed executable names for our target.
+    let is_perm_dir = matches!(**outdir, DirState::Perm(..));
+    let out_dir = outdir.path();
+    let out_dir = if is_multiple_tests && is_perm_dir {
+        // If this a "multiple tests" case and we generate it into a non temporary directory, we
+        // want to put it into the parent instead.
+        out_dir.parent().unwrap_or(out_dir)
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/doctest.rs:387]: It can be improved by only pushing the test id if it's not a multiple test indeed. It cannot be done in `make_test` because we don't know yet if this is a multiple test.

**Comment by GuillaumeGomez** [comment]
Added more regex replacements. Let's see if I didn't miss one...

**Comment by GuillaumeGomez** [comment]
Seems like my changes to the `main` function detection got a bit too much improved. Gonna check that tomorrow.

**Comment by GuillaumeGomez** [comment]
Finally fixed it!

@notriddle: Thanks for finding out about the broken code cases. Added a test case as well. It's now ready for another review round (thanks in advance! :3 ).

**Comment by GuillaumeGomez** [comment]
Fixed merge conflict.

**Comment by camelid** [comment]
This seems like a good idea in general, but the example given with `static mut` concerns me. For the whole time they've been around, doctests have been completely isolated from each other, and now there is the potential for "spooky action at a distance" as @notriddle mentioned. Ideally we could automatically detect doctests that (indirectly) used statics and either run them separately or emit a warning that `standalone` should be added. Maybe it'd be too difficult to implement or risk producing false positives depending on how stdlib types are implemented, etc. Thoughts?

**Comment by camelid** [comment]
If we do consider the static-detection approach, it could potentially be implemented by creating a MIR query `uses_global_mut_state` or similar that would recursively check if the associated MIR body read or mutated a `static mut` or called a function that did the same.

**Comment by GuillaumeGomez** [comment]
I think it'd be a nice improvement but I don't think it should be a part of this first iteration as I'm pretty sure there are cases where using statics in this context is perfectly fine and we'd need to be go through a lot of code to check in which cases it's ok or not.

**Comment by camelid** [comment]
Fair point. I guess this change is reasonable because it matches the way normal unit tests work (all compiled into one binary). Should we consider running all compatible doctests in a single binary in pre-2024 editions, then run those tests individually too, and check if the result didn't match the results of the individual tests (i.e., combined failed but all individual passed, or combined passed but at least one individual failed)? The challenge of course is that it'd cause a further slowdown on pre-2024 tests. Just a thought though to help migration to the new edition behavior.

Also curious what some other people on the team think about this change in behavior.

**Comment by GuillaumeGomez** [comment]
> Fair point. I guess this change is reasonable because it matches the way normal unit tests work (all compiled into one binary). Should we consider running all compatible doctests in a single binary in pre-2024 editions, then run those tests individually too, and check if the result didn't match the results of the individual tests (i.e., combined failed but all individual passed, or combined passed but at least one individual failed)? The challenge of course is that it'd cause a further slowdown on pre-2024 tests. Just a thought though to help migration to the new edition behavior.

I'm not sure to understand what you have in mind in this case. Is it to uncover bugs we might have missed maybe?

> Also curious what some other people on the team think about this change in behavior.

That's why we'll have an FCP in any case. :)

**Comment by camelid** [comment]
> I'm not sure to understand what you have in mind in this case. Is it to uncover bugs we might have missed maybe?

No, more to detect problems with user code. E.g. if doctests in someone's crate do something like depend on the same mutable global state, we can warn them that this will break in Rust 2024.

**Comment by GuillaumeGomez** [comment]
I see. I'm not sure it's a good idea though. Maybe adding an option to force this new behaviour on non-2024 edition runs could be a way for users to give it a try?

**Comment by camelid** [comment]
Yeah, we should probably focus on that for now. Could also consider a crater run.

**Comment by GuillaumeGomez** [comment]
I'm still hoping we can make it before the 2024 edition is done. Let's hope I'm not too optimistic here. ^^'

**Review by camelid — CHANGES_REQUESTED** [review]
Left a few comments about the docs. In terms of the code itself, I believe we should refactor the doctest code before making this change. It's hard to follow what's happening since test collection, processing, and running are all mixed together—not because of this PR, rather due to the existing code. There are also other issues like functions with tons of arguments, which is improved somewhat here but should ideally be done in a separate PR that doesn't change behavior.

I'm currently working on such a refactor and will open a PR soon. Just to be clear, my goal is not to overhaul the test running system but just try to organize and decouple the code better.

File: tests/run-make/doctests-merge/rmake.rs [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    should_succeed: bool,
+    dep: &Path,
+) {
+    let mut cmd = rustdoc();
+
+    cmd.input(input_file)
+        .arg("--test")
+        .arg("-Zunstable-options")
+        .arg("--edition")
+        .arg(edition)
```
Comment (jieyouxu) [inline tests/run-make/doctests-merge/rmake.rs:19]: Suggestion: add an `edition()` helper to `Rustdoc` just like the one on `Rustc`.

File: tests/run-make/doctests-merge/rmake.rs [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        .arg("--test-args=--test-threads=1")
+        .arg("--extern")
+        .arg(format!("foo={}", dep.display()))
```
Comment (jieyouxu) [inline tests/run-make/doctests-merge/rmake.rs:22]: Suggestion: add an `extern_()` helper like the one on `Rustc`.

**Comment by camelid** [comment]
> I'd consider this PR blocked on https://github.com/rust-lang/rust/pull/125798.

Update: it's merged now.

**Comment by GuillaumeGomez** [comment]
On it!

**Comment by GuillaumeGomez** [comment]
Closing in favour of https://github.com/rust-lang/rust/pull/126245.
